src/metaxy/metadata_store/base.py

```diff
@@ -480,30 +483,65 @@ def write_metadata(
         if not is_system_table:
             self._validate_project_write(feature)
 
-        # Convert Narwhals to Polars if needed
-        if isinstance(df, nw.DataFrame):
-            df = df.to_polars()
-        # nw.DataFrame also matches as DataFrame in some contexts, ensure it's Polars
-        if not isinstance(df, pl.DataFrame):
-            # Must be some other type - shouldn't happen but handle defensively
-            if hasattr(df, "to_polars"):
-                df = df.to_polars()
-            elif hasattr(df, "to_pandas"):
-                df = pl.from_pandas(df.to_pandas())
-            else:
-                raise TypeError(f"Cannot convert {type(df)} to Polars DataFrame")
+        # Convert Narwhals to native Polars (both lazy and eager)
+        if isinstance(df, (nw.LazyFrame, nw.DataFrame)):
+            df = df.to_native()  # type: ignore[assignment]
+
+        if pa is not None and isinstance(df, pa.Table):
+            # Streaming paths may yield Arrow tables; normalize to Polars.
+            from typing import cast
+
+            df = cast(pl.DataFrame, pl.from_arrow(df))
+
+        # Type narrowing: at this point df must be pl.DataFrame or pl.LazyFrame
+        if not isinstance(df, (pl.DataFrame, pl.LazyFrame)):
+            raise TypeError(
+                f"Expected Polars DataFrame or LazyFrame after conversion, got {type(df)}"
+            )
 
-        # For system tables, write directly without feature_version tracking
+        # For system tables, collect lazy frames and write directly
         if is_system_table:
+            if isinstance(df, pl.LazyFrame):
+                df = df.collect()
```

Comment:
I'm not sure, why exactly is this placed here? shouldn't `_write_metadata_impl` handle collecting if it's really needed? Collection should be deferred as long as possible

Reply:
fixed